### PR TITLE
Clean architecture first attempt

### PR DESCRIPTION
**Pull request by exRomeo**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -29,11 +34,11 @@ android {
         }
     }
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_17
```
Comment (gemy411) [inline app/build.gradle:37]: Is there a reason for JDK level increase?

File: app/src/main/java/com/trianglz/weatherapp/data/apiservice/apininja/ApiNinja.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package com.trianglz.weatherapp.data.apiservice.apininja
+
+import com.trianglz.weatherapp.data.models.city.CityDto
+import com.trianglz.weatherapp.data.models.weather.WeatherDto
+import retrofit2.http.GET
+import retrofit2.http.Header
+import retrofit2.http.Query
+
+interface ApiNinja {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/data/apiservice/apininja/ApiNinja.kt]: It is preferred to have a data flow for each data type.

File: app/src/main/java/com/trianglz/weatherapp/data/mappers/weather/WeatherMappers.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.trianglz.weatherapp.data.mappers.weather
+
+import com.trianglz.weatherapp.data.models.city.CityDto
+import com.trianglz.weatherapp.data.models.weather.WeatherDto
+import com.trianglz.weatherapp.domain.models.weather.Weather
+
+fun WeatherDto.toWeather(city: CityDto): Weather =
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/data/mappers/weather/WeatherMappers.kt]: In our projects, domain models have the name domain model. Like WeatherDomainModel.

Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/data/mappers/weather/WeatherMappers.kt]: And retrofit return model will be WeatherDataModel

Comment (gemy411) [reply app/src/main/java/com/trianglz/weatherapp/data/mappers/weather/WeatherMappers.kt]: It's a convention not right or wrong.

File: app/src/main/java/com/trianglz/weatherapp/data/remotesource/IRemoteDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package com.trianglz.weatherapp.data.remotesource
+
+import com.trianglz.weatherapp.data.models.city.CityDto
+import com.trianglz.weatherapp.data.models.country.CountryDto
+import com.trianglz.weatherapp.data.models.weather.WeatherDto
+
+interface IRemoteDataSource {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/data/remotesource/IRemoteDataSource.kt]: Our convention also, is to have the interface named like CitiesRemoteSource. and it's implementer will be CitiesRemoteSourceImpl

File: app/src/main/java/com/trianglz/weatherapp/data/remotesource/IRemoteDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    suspend fun getCountries(
+        countryName: String
+    ): List<CountryDto>
+
+    suspend fun getCities(
+        countryCode: String,
+        limit: Int
+    ): List<CityDto>
+
+    suspend fun getWeather(
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/data/remotesource/IRemoteDataSource.kt]: Don't forget to separate data flow for each type: cities, countries and weather.

File: app/src/main/java/com/trianglz/weatherapp/di/modules/ApiModule.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+class ApiModule {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/di/modules/ApiModule.kt]: A DI module will be provided for each data type: cities, countries and weather. To achieve separation.

File: app/src/main/java/com/trianglz/weatherapp/di/modules/ApiModule.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    @Provides
+    @Named("ApiKey")
+    @Singleton
+    fun provideApiKey(): String {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/di/modules/ApiModule.kt]: Api key is a top level variable shouldn't be included in the minor modules. I'm thinking something application level module.

File: app/src/main/java/com/trianglz/weatherapp/di/modules/UseCaseModule.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+abstract class UseCaseModule {
+
+    @Binds
+    @Singleton
+    abstract fun bindCountrySearchUseCase(countrySearchUseCase: CountrySearchUseCase): ICountrySearchUseCase
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/di/modules/UseCaseModule.kt]: On the same scheme as above. let's say we have the cities module: 
Our DI module will handle injection of the use case it's repos and it's remote sources.

File: app/src/main/java/com/trianglz/weatherapp/di/modules/WeatherAppModule.kt [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Binds
+    @Singleton
+    abstract fun bindsRemoteDataSource(remoteDataSource: RemoteDataSource): IRemoteDataSource
+
+
+
+    @Binds
+    @Singleton
+    abstract fun bindsExceptionHandler(exceptionHandler: ExceptionHandler): IExceptionHandler
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/di/modules/WeatherAppModule.kt]: Please clarify to me why the exception handler is injected.

File: app/src/main/java/com/trianglz/weatherapp/domain/models/weather/Weather.kt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+) {
+    fun getWeatherDescription(): String = when {
+        currentTemperature >= 30.0 -> "Hot."
+        currentTemperature >= 20.0 -> "Warm."
+        currentTemperature >= 10.0 -> "Cool."
+        currentTemperature >= 0.0 -> "Freezing!"
+        else -> "Stay indoors."
+    }
+
+    fun getWeatherIcon(): Int {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/models/weather/Weather.kt]: Weather icon is UI related property. It is preferred to have a separate model for the UI related aspects.

File: app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/CountrySearchUseCase.kt [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+import javax.inject.Inject
+
+
+class CountrySearchUseCase @Inject constructor(
+    private val repository: IRepository,
+    private val utilityManager: IUtilityManager
+) : ICountrySearchUseCase {
+    override suspend fun getCountries(
+        countryName: String
+    ): Result<List<Country>> = coroutineScope {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/CountrySearchUseCase.kt]: Kotlin already provides a function for handling such matter. runCatching { }. can be used to create a result object out of the box.

File: app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/ICountrySearchUseCase.kt [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.trianglz.weatherapp.domain.usecases.countrysearch
+
+import com.trianglz.weatherapp.domain.models.country.Country
+
+interface ICountrySearchUseCase {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/ICountrySearchUseCase.kt]: Please clarify the need to have an interface for a use case.

File: app/src/main/java/com/trianglz/weatherapp/domain/usecases/getweatherdata/WeatherDataUseCase.kt [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+class WeatherDataUseCase @Inject constructor(
+    private val repository: IRepository,
+    private val utilityManager: IUtilityManager
+) : IWeatherDataUseCase {
+
+    override suspend fun getWeatherData(
+        countryCode: String,
+        limit: Int
+    ): Result<List<Weather>> = coroutineScope {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/usecases/getweatherdata/WeatherDataUseCase.kt]: Very neat. But, what I'd do is have a use case for fetching the weather data for each city. And from the view model I'd loop and fetch multiple weather datas for each city.

File: app/src/main/java/com/trianglz/weatherapp/domain/utils/exceptionhandler/ExceptionHandler.kt [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+class ExceptionHandler @Inject constructor() : IExceptionHandler {
+    override fun handleException(throwable: Throwable): String = when (throwable) {
+        is HttpException -> handleHttpException(exception = throwable)
+        is SocketTimeoutException -> "Request timed out. Please check your internet connection and try again."
+        is IOException -> "Connection Error, Please check your internet connection."
+        else -> "An error occurred. Please try again later."
+    }
+
+    private fun handleHttpException(exception: HttpException): String = when (exception.code()) {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/utils/exceptionhandler/ExceptionHandler.kt]: Great approach!!

**Review by gemy411 — COMMENTED** [review]
Amazing work so far! Keep it up!

File: app/build.gradle (hunk 1 above)
Comment (exRomeo) [reply app/build.gradle:37]: latest compose version needed a more recent compose compiler which needed a more recent Kotlin version which asked for a more recent jdk  so I just kept choosing the latest versions to get over the "not compatible versions" warnings and errors

short version: a series of incompatible versions led me to choose the latest jdk

File: app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/ICountrySearchUseCase.kt (hunk 12 above)
Comment (exRomeo) [reply app/src/main/java/com/trianglz/weatherapp/domain/usecases/countrysearch/ICountrySearchUseCase.kt]: to make testing the view model easier

File: app/src/main/java/com/trianglz/weatherapp/data/apiservice/apininja/ApiNinja.kt [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+    @GET("city/")
+    suspend fun getCities(
+        @Header("X-Api-Key") apiKey: String,
+        @Query("country") countryCode: String,
+        @Query("limit") limit: Int
+    ): List<CityDataModel>
+
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/data/apiservice/apininja/ApiNinja.kt]: Still having two function in the API ninja interface. On a larger scale this won't support extension and furhter mutations.

File: app/src/main/java/com/trianglz/weatherapp/presentation/exceptionresolver/ExceptionResolverImpl.kt [hunk 16]
```diff
@@ -0,0 +1,43 @@
+package com.trianglz.weatherapp.presentation.exceptionresolver
+
+import android.content.res.Resources
+import com.trianglz.weatherapp.R
+import retrofit2.HttpException
+import java.io.IOException
+import java.net.SocketTimeoutException
+import javax.inject.Inject
+
+class ExceptionResolverImpl @Inject constructor(private val resourceManager: Resources) :
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/presentation/exceptionresolver/ExceptionResolverImpl.kt:10]: The pinnacle of human engineering XD 
Love it

File: app/src/main/java/com/trianglz/weatherapp/presentation/exceptionresolver/ExceptionResolverImpl.kt [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    ExceptionResolver {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/presentation/exceptionresolver/ExceptionResolverImpl.kt:11]: If you want to achieve better separation. Here is a tip: 
It's a good decision to have this in the presentation layer. 
But talking about HTTP errors in UI layers isn't optimum. 
The ideal flow of thought would be that, the domain shoots for the data layer, while data layer errors are mapped to domain layer errors, and those domain layer errors are mapped to UI String in the UI layer. 
It's not necessary but it's how the chain of though in the clean arch works. Still, this  code works perfectly and still very modular and maintainable. So, good job!!

File: app/src/main/java/com/trianglz/weatherapp/domain/repository/city/CitiesRepository.kt [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package com.trianglz.weatherapp.domain.repository.city
+
+import com.trianglz.weatherapp.data.models.city.CityDataModel
+
+interface CitiesRepository {
+
+    suspend fun getCities(
+        countryCode: String,
+        limit: Int
+    ): List<CityDataModel>
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/repository/city/CitiesRepository.kt:10]: This repository declaration is a domain specidifc object. So, referencing data models isn't recommended.

File: app/src/main/java/com/trianglz/weatherapp/di/modules/city/CityModule.kt [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+import com.trianglz.weatherapp.domain.usecases.fetchcities.FetchCitiesUseCaseImpl
+import dagger.Binds
+import dagger.Module
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+abstract class CityModule {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/di/modules/city/CityModule.kt:21]: perfect execution

File: app/src/main/java/com/trianglz/weatherapp/domain/usecases/fetchcities/FetchCitiesUseCase.kt [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+
+    /**
+     * [getCities] just calls the method from the repository and wraps it in [runCatching] block
+     * which returns one of two case success with the received data or failure with the Exception
+     * */
+
+    suspend fun getCities(
+        countryCode: String,
+        limit: Int
+    ): Result<List<CityDataModel>>
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/domain/usecases/fetchcities/FetchCitiesUseCase.kt:15]: Data models are data layer specific. 
Here we use domain models.

File: app/src/main/java/com/trianglz/weatherapp/presentation/ui/home/HomeViewModel.kt [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+     * then checks if the list is empty it sets the ui state to failure with message explaining that no weather data is found
+     * if the list has data it sets the ui state to success and displays the weather data
+     * */
+
+    private suspend fun getWeather(
+        cityList: List<CityDataModel>
+    ) = coroutineScope {
+        val weatherResult =
+            cityList.map { city ->
+                async {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/presentation/ui/home/HomeViewModel.kt:186]: Very clean! Having async the return of map. Love it.

File: app/src/main/java/com/trianglz/weatherapp/presentation/extensions/ModifierExtensions.kt [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.platform.LocalConfiguration
+import androidx.compose.ui.unit.dp
+
+/**
+ * @param ratio is the percentage of screen's height the component's maximum height will be
+ * */
+
+fun Modifier.maxHeightScreenRatio(@FloatRange(0.2, 1.0) ratio: Float) = composed {
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/presentation/extensions/ModifierExtensions.kt:14]: Good convention.

File: app/src/main/java/com/trianglz/weatherapp/presentation/ui/components/WeatherCard.kt [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+import androidx.compose.ui.unit.dp
+import com.trianglz.weatherapp.R
+import com.trianglz.weatherapp.data.models.weather.WeatherDataModel
+import com.trianglz.weatherapp.presentation.ui.theme.BackgroundGradient
+import com.trianglz.weatherapp.presentation.ui.theme.PurpleIndigoGradient
+import com.trianglz.weatherapp.presentation.ui.theme.WeatherAppTheme
+import com.trianglz.weatherapp.presentation.ui.theme.lavender
+
+@Composable
+fun WeatherCard(
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/presentation/ui/components/WeatherCard.kt:39]: Why does the weather card do not have a data class for it's state like the search bar?

File: app/src/main/java/com/trianglz/weatherapp/presentation/searchbarstate/SearchBarState.kt [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+) = remember(initialState) {
+    initialState
+}
+
+@Composable
+fun rememberSearchState(
+    placeHolder: String = "",
+    noResultMessage: String = "",
+    status: SearchBarStatus = SearchBarStatus.Idle,
+    result: List<CountryDomainModel> = emptyList(),
```
Comment (gemy411) [inline app/src/main/java/com/trianglz/weatherapp/presentation/searchbarstate/SearchBarState.kt]: The search bar should be decoupled from the list of results (Countries)

**Review by gemy411 — CHANGES_REQUESTED** [review]
Keep up the great work! This app is almost production ready!

**Pull request merged**